crates/cext/src/transpiler/passes/commutative_cancellation.rs

```diff
@@ -0,0 +1,127 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2025
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use crate::exit_codes::ExitCode;
+use crate::pointers::{const_ptr_as_ref, mut_ptr_as_ref};
+
+use qiskit_circuit::circuit_data::CircuitData;
+use qiskit_circuit::converters::dag_to_circuit;
+use qiskit_circuit::dag_circuit::DAGCircuit;
+use qiskit_transpiler::commutation_checker::get_standard_commutation_checker;
+use qiskit_transpiler::passes::cancel_commutations;
+use qiskit_transpiler::target::Target;
+
+/// @ingroup QkTranspilerPasses
+/// Run the CommutativeCancellation transpiler pass on a circuit.
+///
+/// This pass cancels the redundant (self-adjoint) gates through commutation relations.
+///
+/// @param circuit A pointer to the circuit to run CommutativeCancellation on. This circuit
+/// pointer to will be updated with the modified circuit if the pass is able to remove any gates.
+/// @param target The target for the CommutativeCancellation pass. The target is used to specify
+/// as fallback Z rotation gate to use for cancellations if one can't be identified in the circuit.
+/// This can be a null pointer which indicates there isn't a target to use.
```

Comment:
Not saying this behavior makes sense but this seems to be what it does 🙃 
```suggestion
/// @param target This pass will attempt to accumulate all Z rotations into either
/// an RZ, P or U1 gate, depending on which is already used in the circuit. If none
/// is present in the circuit, this (optional) target argument is used as fallback to 
/// decide which gate to use. If none of RZ, P or U1 are in the circuit or the target, 
/// single-qubit Z rotations will not be optimized.
```

Reply:
So here is my interpretation of what the pass _actually_ does:

The pass does two things:

* it removes consecutive pairs of (self-inverse) CX, CY, CZ, H or Y gates,
* it combines consecutive RZ-rotations (which include P, U1, RZ, Z, S and T gates), into either an RZ, P or U1 gate, and combines consecutive RX-rotations (which include X and RX) into an RX gates.

where "consecutive" takes the commutativity into account. All the lists above are hard-coded (first in the python code and second in the rust code).

The type of the gate we use to combine Z-rotations (i.e. RZ, P or U1) depends on which gates are already used in the circuit. As Julien explained: If none is present in the circuit, this (optional) target argument is used as fallback to  decide which gate to use. If none of RZ, P or U1 are in the circuit or the target,  single-qubit Z rotations will not be optimized.

---

Notes: I do not understand why we don't combine all Z-rotations into RZ, but support the option of using either RZ, P or U1, and do the complicated analysis that we do. Can we just use RZ?

There is currently a bug that will get fixed in #14956.

As discussed with Julien, the pass also misses Sdg and Tdg gates. These would be very simple to add but this is technically a new feature, so we can't add this for 2.2. Or can we?

Reply:
> Notes: I do not understand why we don't combine all Z-rotations into RZ, but support the option of using either RZ, P or U1, and do the complicated analysis that we do. Can we just use RZ?

RZ might not be in the basis or the target that the user supplies, so I don't think we can just use RZ -- unless we add a translation layer. I wouldn't change any of this now until we rework the commutative optimization into a single pass, there we can revisit this properly.

Reply:
Also note that this doc is only for the `target` attribute, which is only used to determine the Z-rotation type; the general description should go into the docstring or just the Python docs as we generally refer to the Python docs for the C passes right now

Reply:
FWIW, I'm not too worried about maintaining basis gates here, the pass actually doesn't do it correctly right now anyway. It makes some bad assumptions about the homogeneity of the target, just having an rz gate in the target doesn't mean it's valid on every qubit. So if we have a potential path to better optimization I wouldn't worry about ensuring the gates are correct for the target we already assume optimization passes may not do it correct in the `optimization` stage which is why we run the basis translator at the end of each loop. We also run this pass in the init stage and the gates we use don't matter there.